Automattic/wp-calypso
# Color Schemes: Add color schemes via CSS custom properties

This PR is part of a series of PRs aiming to provide support for Color Schemes in Calypso.
This PR focuses on adding color scheme support via CSS custom properties.

![ezgif com-video-to-gif 3](https://user-images.githubusercontent.com/1562646/29487542-12f3aa90-84fb-11e7-937b-83e6bb4f69fb.gif)

## General

### Notable additions in this PR:
- This PR adds the plugin `postcss-custom-properties` to the postcss build step
- The options for `autoprefixer` and `postcss-custom-properties` are combined in a dedicated `postcss.config.json` config file in the project root
- A new file `stylesheets/shared/_color-schemes.scss` that contains the color schemes
- A `dark` and `light` color scheme are added and demonstrated on the Masterbar

### Considerations

- enable instant preview of the selected color scheme
- easily switch between color schemes via a single class
- define the styles for a new theme in a single location
- neatly integrates into the current CSS architecture
- minimal impact on overall CSS file size and build complexity
- keep default theme exactly as it is
- treat it as a progressive enhancement (current theme remains the default)

For a more detailed description of the feature, the chosen approach and its advantages feel free to check out my post on the subject: `p4TIVU-75d-p2` (internal reference). 

Hat tip to @folletto for a good idea that resulted in the solution presented here, which offers an even better developer experience than the initially suggested one. :)

### How to test
- Visit https://calypso.live/?branch=add/color-schemes/add-style-portion or test locally.
- Navigate to the `Accounts Settings` page and scroll down to the `Admin Color Scheme` section. 
- Depending on whether the browser supports CSS custom properties, the Admin Color Scheme Picker will be present or not (for browser support see http://caniuse.com/#feat=css-variables)
- Select a color scheme to change the appearance of the Masterbar
- Save to persist or reload without saving to discard your selection


### Suggestions as to what to test
- Do the Color Schemes work in all browsers that support custom properties ([see browser support](http://caniuse.com/#feat=css-variables))?
- Is the Color Scheme Selector on the Account Settings page hidden when viewing on a browser that doesn't support custom properties (aka does the feature detection work)?
- Is the responsive behavior sufficient?
- Is the selection being persisted when saved? Is it being discarded when reloading without saving?


### Notes

This PR currently compares its changes to PR https://github.com/Automattic/wp-calypso/pull/17307. This is so the changes can be easily identified. The following PRs (in chronological order) need to be merged prior to this PR:

PR https://github.com/Automattic/wp-calypso/pull/17157: `Color Schemes: Add new form components as basis for color scheme switcher`
PR https://github.com/Automattic/wp-calypso/pull/17200: `Color Schemes: Add Color Scheme Picker`
PR https://github.com/Automattic/wp-calypso/pull/17259: `Color Schemes: Add Color Scheme Picker to Account Settings`
PR https://github.com/Automattic/wp-calypso/pull/17307: `Color Schemes: Add selected Color Scheme to Layout`
PR https://github.com/Automattic/wp-calypso/pull/17310: `Color Schemes: Add Feature Detection`

This PR will be pointed at `master` once all of the above have been merged.

## Design Review

I'd like to use this PR to also get some design review on the visual elements introduced in this and the PRs mentioned above. I'm thus also posting some additional screenshots here of the visual elements that are being introduced by the Color Scheme PRs.

#### Form Components: Form Radio with Thumbnail & Form Radios Bar
`devdocs/design/form-fields` – first introduced in PR https://github.com/Automattic/wp-calypso/pull/17157

Using several `FormRadioWithThumbnail` components will list them vertically. For a nicer visual appearance, they can be wrapped in the `FormRadiosBar` which will display them in a horizontal grid. The `FormRadiosBar` also provides a nicer developer experience for both normal and thumbnail radios as they can now be passed in as an object.

<img width="942" alt="screen shot 2017-08-19 at 17 31 23" src="https://user-images.githubusercontent.com/1562646/29488110-62907610-8504-11e7-9985-488537107358.png">




#### Blocks: Color Scheme Picker 
`devdocs/blocks/color-scheme-picker` – first introduced in PR https://github.com/Automattic/wp-calypso/pull/17200

This is added as a block with its potential future of becoming a per-person per-site selection in mind. That would involve adding more components like e.g. a site selector.

Note: I created the theme previews in Photoshop, that's why they are currently PNGs. I've already noted down converting them to SVG as a future todo. Or if you'd like to help with that... :)

![color-scheme-picker](https://user-images.githubusercontent.com/1562646/29488065-d2acf190-8503-11e7-8ed4-65ed1785d8f6.gif)
 
Thanks for reading! 🚀
